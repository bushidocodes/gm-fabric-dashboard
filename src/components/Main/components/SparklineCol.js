import { spacingScale, contrastColor } from "style/styleFunctions";
import {
  BORDER_RADIUS_BASE,
  COLOR_CONTENT_BACKGROUND
} from "style/styleVariables";

import TableCol from "./TableCol";

const SparklineCol = TableCol.extend`
  padding: 0;
  height: ${spacingScale(3)};
  position: relative;
  display: flex;

  svg {
    flex: 1 1 100%;
    border-radius: ${BORDER_RADIUS_BASE};
    background-color: rgba(144, 144, 144, 0.1);
  }
`;

export default SparklineCol;
