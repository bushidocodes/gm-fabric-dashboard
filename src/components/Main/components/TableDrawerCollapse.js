import styled from "styled-components";
import Collapse from "react-collapse";

import { spacingScale } from "style/styleFunctions";

const TableDrawerCollapse = styled(Collapse)`
  cursor: default;
  flex: 0 0 100%;
  min-height: 0;
  overflow: hidden;
  padding: 0;
  white-space: normal;

  > div:first-child {
    box-sizing: border-box;
    cursor: default;
    height: auto;
    margin: 0 ${spacingScale(2)};
    padding: ${spacingScale(2)} 0;
    text-align: left;
    white-space: normal;
  }
`;

export default TableDrawerCollapse;
