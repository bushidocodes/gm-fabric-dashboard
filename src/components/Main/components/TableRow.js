import styled from "styled-components";
import { PropTypes } from "prop-types";

import {
  spacingScale,
  contrastColor,
  rowChildSpacing
} from "style/styleFunctions";
import {
  COLOR_CONTENT_BACKGROUND,
  TABLE_HOVER,
  TABLE_BORDER
} from "style/styleVariables";

const TableRow = styled.li.attrs({
  tabIndex: 0
})`
  align-items: center;
  background-color: ${props =>
    props.isOpen
      ? contrastColor(COLOR_CONTENT_BACKGROUND, 0.04).string()
      : contrastColor(COLOR_CONTENT_BACKGROUND, 0).string()};
  box-shadow: inset 0 -1px ${TABLE_BORDER};
  display: flex;
  flex-wrap: wrap;
  min-height: ${spacingScale(4.5)};
  width: 100%;
  cursor: ${props => (props.selectable ? `pointer` : `default`)};

  &:hover {
    ${props => props.selectable && `background-color: ${TABLE_HOVER}; `};
  }
  &:focus {
    outline: -webkit-focus-ring-color auto 5px;
  }

  &:focus:active:hover {
    outline: 0;
  }

  > * {
    ${rowChildSpacing()};
  }
`;

TableRow.propTypes = {
  open: PropTypes.bool
};

export default TableRow;
