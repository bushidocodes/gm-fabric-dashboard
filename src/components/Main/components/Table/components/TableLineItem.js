import _ from "lodash";
import { PropTypes } from "prop-types";
import React, { Component } from "react";
import {
  Sparklines,
  SparklinesLine,
  SparklinesReferenceLine
} from "react-sparklines";

import GMLineChart from "components/Main/components/GMLineChart";
import TableRow from "components/Main/components/TableRow";
import TableCol from "components/Main/components/TableCol";
import TableColVizBar from "components/Main/components/TableColVizBar";
import SparklineCol from "components/Main/components/SparklineCol";
import TableDrawerCollapse from "components/Main/components/TableDrawerCollapse";
import VizBar from "components/Main/components/VizBar";
import VizFill from "components/Main/components/VizFill";
import Badge from "components/Main/scenes/InstanceView/components/Badge";
import Icon from "components/Icon";
import Glyph from "components/Glyphs/index";

import styled from "styled-components";

// this extra flex container is necessary to truncate route name in chrome browser
const FlexParent = styled.div`
  display: flex;
  > div {
    margin-top: 5px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    min-width: 0;
  }
`;

/**
 * A row of data in Functions and Routes table
 * @export
 * @class TableLineItem
 * @extends {Component}
 */
export default class TableLineItem extends Component {
  static propTypes = {
    errorPercent: PropTypes.string.isRequired,
    item: PropTypes.string.isRequired,
    latency50: PropTypes.number.isRequired,
    latency99: PropTypes.number.isRequired,
    relativeReqPercent: PropTypes.number,
    requests: PropTypes.number,
    requestsPerSecond_dygraph: PropTypes.array.isRequired,
    requestsPerSecond_sparkline: PropTypes.array.isRequired,
    verb: PropTypes.string
  };

  state = {
    isOpen: false
  };

  // In IE10, IE11, and Edge broswer, SVG elements are added to tab order by default and tabIndex is ignored.
  // setting focusable: "false" attribute to svg will make svg unfocusable.
  // Sparkline module returns its own svg element so the attribute is set here instead of being
  // set directly on svg.
  componentDidMount() {
    const SparklineSVG = document.querySelectorAll("svg[preserveAspectRatio]");
    if (!_.isEmpty(SparklineSVG))
      _.forEach(SparklineSVG, svg => svg.setAttribute("focusable", "false"));
  }

  blurTableRow = e => {
    // this is done to search up the DOM tree to find table row and take away its focus to prevent outline on click while preserving tabbing outline
    let node = e.target;
    while (
      typeof node.className !== "string" ||
      node.className.indexOf("TableRow") !== 0
    ) {
      node = node.parentNode;
    }
    node.blur();
  };

  toggleDrawer = e => {
    if (e) {
      this.blurTableRow(e);
    }

    this.setState({ isOpen: !this.state.isOpen });
  };

  render() {
    return (
      <TableRow
        selectable={this.state.isOpen}
        onClick={evt => this.toggleDrawer(evt)}
        onKeyDown={evt => {
          if (evt.keyCode === 13) {
            evt.preventDefault();
            this.toggleDrawer();
          }
        }}
        role="link"
      >
        <TableColVizBar>
          <FlexParent>
            {this.props.verb && (
              <Badge>
                <Icon>
                  <Glyph name={this.props.verb} />
                </Icon>
                {this.props.verb}
              </Badge>
            )}
            <div>{this.props.item}</div>
          </FlexParent>
          <VizBar>
            <VizFill
              width={this.props.relativeReqPercent}
              colorDegree={Number(this.props.errorPercent)}
            />
          </VizBar>
        </TableColVizBar>

        <SparklineCol>
          <Sparklines
            data={this.props.requestsPerSecond_sparkline}
            height={32}
            preserveAspectRatio="xMaxYMin"
          >
            <SparklinesLine
              style={{
                stroke: "currentColor",
                strokeWidth: 1,
                fill: "none"
              }}
            />
            <SparklinesReferenceLine
              style={{
                stroke: "grey",
                opacity: "0.4"
              }}
              type="mean"
            />
          </Sparklines>
        </SparklineCol>
        <TableCol style={{ textAlign: "right" }}>
          {this.props.requests && this.props.requests.toLocaleString()}
        </TableCol>
        <TableCol
          style={{ textAlign: "right" }}
          errorPercent={this.props.errorPercent}
        >
          {this.props.errorPercent}%
        </TableCol>
        <TableCol
          style={{
            textAlign: "right",
            display: "flex",
            flexDirection: "column"
          }}
        >
          <div>{this.props.latency50}ms</div>
          <div>{this.props.latency99}ms</div>
        </TableCol>
        <TableDrawerCollapse
          isOpened={this.state.isOpen}
          onClick={evt => {
            evt.stopPropagation();
            this.blurTableRow(evt);
          }}
        >
          <GMLineChart
            timeSeries={this.props.requestsPerSecond_dygraph}
            title={"Requests over Time for " + this.props.item}
          />
        </TableDrawerCollapse>
      </TableRow>
    );
  }
}
