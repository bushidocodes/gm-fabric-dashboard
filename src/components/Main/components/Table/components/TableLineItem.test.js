import React from "react";
import { mount, shallow } from "enzyme";
import renderer from "react-test-renderer";

import TableLineItem from "./TableLineItem";
import TableRow from "components/Main/components/TableRow";
import TableCol from "components/Main/components/TableCol";
import TableColVizBar from "components/Main/components/TableColVizBar";
import SparklineCol from "components/Main/components/SparklineCol";
import TableDrawerCollapse from "components/Main/components/TableDrawerCollapse";

let wrapper;

const mockedEvent = {
  target: {
    className: "TableRow",
    blur: () => {}
  }
};

const TableLineItemWithProps = (
  <TableLineItem
    errorPercent={"10.000"}
    errorsCount={32}
    item={"CatalogStream"}
    latency50={445}
    latency99={35}
    relativeReqPercent={20}
    requests={14333}
    requestsPerSecond_dygraph={[0, 1, 5, 30, 56]}
    requestsPerSecond_sparkline={[0, 25, 430, 1256]}
  />
);

describe("<TableLineItem/>", () => {
  beforeEach(() => {
    wrapper = mount(TableLineItemWithProps);
  });

  test("renders styled-components", () => {
    expect(wrapper.find(TableRow).length).toBe(1);
    expect(wrapper.find(TableColVizBar).length).toBe(1);
    expect(wrapper.find(SparklineCol).length).toBe(1);
    expect(wrapper.find(TableCol).length).toBe(3);
    expect(wrapper.find(TableDrawerCollapse).length).toBe(1);
  });

  // use 'shallow' instead of mount for instance tests
  test("should toggle row's open/closed state when row is clicked", () => {
    wrapper = shallow(TableLineItemWithProps);

    const row = wrapper.find(TableRow);
    row.simulate("click", mockedEvent);
    expect(wrapper.state().isOpen).toEqual(true);

    row.simulate("click", mockedEvent);
    expect(wrapper.state().isOpen).toEqual(false);
  });

  test("should toggle drawer's open/closed state when row is clicked", () => {
    wrapper = shallow(TableLineItemWithProps);

    const row = wrapper.find(TableRow);
    row.simulate("click", mockedEvent);
    const drawer = wrapper.find(TableDrawerCollapse);

    expect(wrapper.state().isOpen).toEqual(true);
    expect(drawer.props().isOpened).toBe(true);
  });

  test("matches snapshot", () => {
    const tree = renderer.create(TableLineItemWithProps).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
