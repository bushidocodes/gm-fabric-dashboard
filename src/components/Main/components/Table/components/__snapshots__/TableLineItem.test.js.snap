// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableLineItem/> matches snapshot 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: rgb(255,255,255);
  box-shadow: inset 0 -1px hsl(0,0%,92%);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  min-height: 36px;
  width: 100%;
  cursor: pointer;
}

.c0:hover {
  background-color: hsl(0,0%,98%);
}

.c0:focus {
  outline: -webkit-focus-ring-color auto 5px;
}

.c0:focus:active:hover {
  outline: 0;
}

.c0 > * {
  padding-right: 16px;
  padding-top: 7px;
  padding-bottom: 7px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c0 > *:first-child {
  padding-left: 16px;
}

.c6 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
}

.c7 {
  color: hsl(0,100%,42.5%);
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
}

.c1 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
  -webkit-flex: 1 1 20%;
  -ms-flex: 1 1 20%;
  flex: 1 1 20%;
  position: relative;
}

.c5 {
  -webkit-flex: 1 1 12%;
  -ms-flex: 1 1 12%;
  flex: 1 1 12%;
  text-align: left;
  padding-top: 4px;
  padding-bottom: 4px;
}

.c5 svg {
  height: 24px;
}

.c3 {
  background-color: rgb(230,230,230);
  border-radius: 1px 4px 4px 1px;
  bottom: 0;
  height: 3px;
  left: 0;
  position: absolute;
  right: 16px;
}

.c4 {
  background-color: currentColor;
  border-radius: inherit;
  bottom: 0;
  color: rgb(32,154,38);
  left: 0;
  min-width: 2px;
  position: absolute;
  top: 0;
  width: 20%;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 > div {
  margin-top: 5px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  min-width: 0;
}

<li
  className="c0"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="link"
  tabIndex={0}
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div>
        CatalogStream
      </div>
    </div>
    <div
      className="c3"
    >
      <div
        className="c4"
        width={20}
      />
    </div>
  </div>
  <div
    className="c5"
  >
    <svg
      preserveAspectRatio="xMaxYMin"
      style={undefined}
      viewBox="0 0 240 32"
    >
      <g>
        <circle
          cx={2}
          cy={29.999999999999996}
          onClick={[Function]}
          onMouseEnter={[Function]}
          r={2}
          style={
            Object {
              "fill": "none",
              "fillOpacity": ".1",
              "pointerEvents": "auto",
              "stroke": "currentColor",
              "strokeWidth": "0",
            }
          }
        />
        <circle
          cx={80.66666666666667}
          cy={29.442675159235666}
          onClick={[Function]}
          onMouseEnter={[Function]}
          r={2}
          style={
            Object {
              "fill": "none",
              "fillOpacity": ".1",
              "pointerEvents": "auto",
              "stroke": "currentColor",
              "strokeWidth": "0",
            }
          }
        />
        <circle
          cx={159.33333333333334}
          cy={20.4140127388535}
          onClick={[Function]}
          onMouseEnter={[Function]}
          r={2}
          style={
            Object {
              "fill": "none",
              "fillOpacity": ".1",
              "pointerEvents": "auto",
              "stroke": "currentColor",
              "strokeWidth": "0",
            }
          }
        />
        <circle
          cx={238}
          cy={2}
          onClick={[Function]}
          onMouseEnter={[Function]}
          r={2}
          style={
            Object {
              "fill": "none",
              "fillOpacity": ".1",
              "pointerEvents": "auto",
              "stroke": "currentColor",
              "strokeWidth": "0",
            }
          }
        />
        <polyline
          points="2 29.999999999999996 80.66666666666667 29.442675159235666 159.33333333333334 20.4140127388535 238 2 238 30 2 30 2 29.999999999999996"
          style={
            Object {
              "fill": "none",
              "fillOpacity": ".1",
              "pointerEvents": "auto",
              "stroke": "currentColor",
              "strokeWidth": "0",
            }
          }
        />
        <polyline
          points="2 29.999999999999996 80.66666666666667 29.442675159235666 159.33333333333334 20.4140127388535 238 2"
          style={
            Object {
              "fill": "none",
              "stroke": "currentColor",
              "strokeLinecap": "round",
              "strokeLinejoin": "round",
              "strokeWidth": 1,
            }
          }
        />
      </g>
      <line
        style={
          Object {
            "opacity": "0.4",
            "stroke": "grey",
          }
        }
        x1={2}
        x2={238}
        y1={22.46417197452229}
        y2={22.46417197452229}
      />
    </svg>
  </div>
  <div
    className="c6"
    style={
      Object {
        "textAlign": "right",
      }
    }
  >
    14,333
  </div>
  <div
    className="c7"
    style={
      Object {
        "textAlign": "right",
      }
    }
  >
    10.000
    %
  </div>
  <div
    className="c6"
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "textAlign": "right",
      }
    }
  >
    <div>
      445
      ms
    </div>
    <div>
      35
      ms
    </div>
  </div>
</li>
`;
